### ISSUE TITLE
State transition to void triggered for animated children while still existing in DOM

### ISSUE THREAD

**Issue by georgiee** [angular/angular#27754]
# 🐞 bug report

### Affected Package
The issue is caused by package `@angular/animation` (any version)

### Description
1. There is an array of numbers
2. There is a parent container that generates multiple child items with a ngFor loop using that array.
3. Each item gets its number from the array assigned.
3. Every item has an animation attached to visualize active or inactive states. The parent container has an animation attached to coordinate staggering of the children.
4. When removing an item by modifying the array of numbers the item is removed.
5. This usually works and the item disappears (no animation involved)

When the array contains identical numbers and you remove one of the duplicate numbers you run into a problem. As expected this will cause remaining siblings to move to the left but you will get a list of "dead" components— but only from the animation perspective as the child animation have been transitioned to state void while remaining safe & sound in the DOM. I can reproduce this problem in a small stackblitz where I go into more details by referring to parts of the demo.

## 🔬 Minimal Reproduction
Preview:
https://animation-dom-move-bug.stackblitz.io/

Source:
https://stackblitz.com/edit/animation-dom-move-bug

## 🔥 Exception or Error
My small [stackblitz demo](https://stackblitz.com/edit/animation-dom-move-bug) shows the error and you can reproduce it perfectly:

**Do this:** Click on "remove 2" and you will see what's happening in this gif.

![bug](https://user-images.githubusercontent.com/1701755/50223899-b819f080-039c-11e9-9723-0427bf6377b6.gif)

Yellow stands for state void. You can see a yellow to green animation at the beginning of the gif loop that's the transition (void => active). After clicking the remove button you will see yellow items. I call them dead. Still in DOM but in state void. Just look at the all yellow numbers after you clicked the remove button. Also look into the console, you will see plenty of state transitions from active to void which is wrong.

## 🌍  Your Environment
**Angular Version:**
Latest Angular 7, see the stackblitz demo.

**Anything else relevant?**
Here some related issues I think
#23534
#26078

**Comment by kvetis** [comment]
I don't think that Angular supports duplicate items in `*ngFor` directive without using a `trackBy` function. Without it, it cannot pair reference the items with their elements. I have [added trackBy to your example](https://stackblitz.com/edit/animation-dom-move-bug-scnq7l?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fitem-holder%2Fitem-holder.component.ts,src%2Fapp%2Fitem-holder%2Fitem-holder.component.html) and it mitigates the issue.

**Comment by thePunderWoman** [comment]
This issue was closed as obsolete due to the introduction of the new `animate.enter` and `animate.leave` APIs. The `@angular/animations` package is now deprecated and will be removed in Angular v23. We hope you give the new APIs a try.
